### PR TITLE
Introduce `flintify` helper for "optimal" dispatch on integer and rational inputs

### PR DESCRIPTION
**Pull request by fingolfin**
There are a bunch of FLINT function that come in multiple varieties. E.g. fmpz_add, fmpz_add_ui, fmpz_add_si all take one fmpz and one additional argument of type fmpz / UInt / Int, respectively.

So naturally we define different `+` and `add!` methods in Julia using them. But which method to invoke when adding a `ZZRingElem` and a Julia integer of a type different from `UInt` and `Int`, such as e.g. `UInt32` ? In that case we currently always convert that into a `ZZRingElem`. But that's inefficient, it would be be better to convert that value to a `UInt` or `Int` value and then use the optimized `fmpz_add_ui/_si` method.

Of course this can be done, but getting it right is tedious, repetitive and it is easy to overlook a case.

This is where `FlintInt` comes in: this is defined as `Union{Int, ZZRingElem}`. We then provide constructors which convert any `Integer` to a `FlintInt` in an optimal way (at least for a bunch of built-in integer types).

This then can be used to write optimal dispatch for Integer types like this:

    add!(x::ZRingElem, y::ZRingElem) = ...
    add!(x::ZRingElem, y::Int) = ...
    add!(x::ZRingElem, y::UInt) = ...
    # fallback code
    add!(x::ZRingElem, y::Integer) = add!(x, FlintInt(y))

It also works for types that accept ZZRingElem and Int, but not UInt:

    add!(x::ZPolyRingElem, y::ZPolyRingElem) = ...
    add!(x::ZPolyRingElem, y::ZRingElem) = ...
    add!(x::ZPolyRingElem, y::Int) = ...
    # fallback code
    add!(x::ZPolyRingElem, y::Integer) = add!(x, FlintInt(y))

Of course I might have missed optimal conversion for some `Integer` subtype. But then we can fix this by simply adding another `FlintInt` constructor in a single central place.

Finally, the new type `FlintRat` plays a similar role for FLINT
functions which also take a `QQFieldElem`. While `RationalUnion` is a
counterpart to `IntegerUnion`.

Note that there are tons more places that could use `FlintInt`, in particular I'd like to use it to make more generic `add!` methods for e.g. `ZZRingElem` and many more (see also [this comment thread](https://github.com/Nemocas/Nemo.jl/pull/1866#discussion_r1775650686) started by @lgoettgens).

But before I invest more work into this, I'd like to hear what others think about this approach. If we decide against it, no point in wasting more effort on it.

### CODE REVIEW

File: src/flint/FlintTypes.jl [hunk 1]
```diff
@@ -6453,8 +6457,79 @@ end
...
+to get optimal dispatch.
+
+This also works for Nemo types that also have special handlers for `UInt`,
+as their method for `b::UInt` takes precedence over the fallback method.
+"""
+const FlintInt = Union{Int, ZZRingElem}
+
+FlintInt(x::ZZRingElem) = x
+FlintInt(x::Integer) = ZZRingElem(x)::ZZRingElem
+if Int === Int64
```
Comment (lgoettgens) [inline src/flint/FlintTypes.jl:6512]:
```suggestion
@static if Int === Int64
```

**Review by lgoettgens — APPROVED** [review]
I really like this idea (but that was probably expected from the other thread). But let's hear for the other opinions.

**Comment by thofma** [comment]
I like the idea, but I am not sure that overloading `(::Union{...})(...)` is the best approach. I have the feeling  this makes things more brittle, prone to invalidations etc. How about just adding a small `flintification` that can be used instead?

**Comment by lgoettgens** [comment]
> I like the idea, but I am not sure that overloading `(::Union{...})(...)` is the best approach. I have the feeling this makes things more brittle, prone to invalidations etc. How about just adding a small `flintification` method that can be used instead?

having a function `flintify(::IntegerUnion) -> FlintInt` that does the conversion, or even `flint_int(::IntegerUnion) -> FlintInt` should work equally well. works for me

**Comment by fingolfin** [comment]
Yes I agree with @thofma's point. I'll rewrite this to use a function `flintify` (in case someone has a better idea: changing the name later will of course be trivial)

**Comment by fingolfin** [comment]
Renaming is done now

**Pull request merged**